lib/spack/docs/build_systems/gopackage.rst

```diff
@@ -0,0 +1,251 @@
+.. Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+   Spack Project Developers. See the top-level COPYRIGHT file for details.
+
+   SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+.. _gopackage:
+
+-----------
+GoPackage
+-----------
+
+There are a couple of things to be aware of about applications built
+with Go:
+
+* The ``go`` command includes more build/install functionality than
+  C's ``cc`` but not as much e.g. Python's ``pip``.
+
+  As a result, Go-based projects use a variety of build strategies:
+  simple invocations of ``go build``, custom go "scripts" that
+  leverage ``go run``, Makefiles, and/or *etc...*.
+
+* Go's dependency management story has evolved from an initial
+  do-it-yourself attitude through a small number of commonly used
+  tools to an officially supported approach (Go "modules").  The Go
+  team provides an `introduction to Go modules
+  <https://blog.golang.org/using-go-modules>`_ and maintains a site
+  that tracks `the gory details
+  <https://github.com/golang/go/wiki/Modules>`_.
+
+  Go modules are the future, ``GOPATH`` will likely be deprecated in
+  Go 1.14.
+
+The best approach to packaging a Go-based project for Spack depends on
+the details of the project.  ``GoPackage`` is a good fit for simple
+situations but another package, e.g. ``MakefilePackage``, might be
```

Comment:
Would it generally be advantageous to combine the logic from `GoPackage` and `MakefilePackage`? Is the issue with doing this the possibility of MRO conflicts?

If yes/yes: I could think about how to combine `GoPackage` with other build systems.

Reply:
I don't think so, at least immediately.  If something uses a Makefile, than adding dependency on a go compiler, declaring resources via `module2tuples` (or by hand if you're a masochist), and using `MakefilePackage` seems like a nice, simple strategy.  I'm open to something more complicated but don't see any need so far.  At some point down the road it might also make sense to make the go compiler "virtual", satisfiable via `go` or `gccgo` or `tinygo` or one of the LLVM-based compilers.  But not until there's a real use case.

FWIW, here's the current (including this PR) breakdown of build techniques:

| APP           | STRATEGY                  |
|---------------|---------------------------|
| hugo          | go build                  |
| rclone        | go build                  |
| lazygit       | go build                  |
| docui         | go build                  |
| go-md2man     | go build                  |
| lazydocker    | go build                  |
| modules2tuple | go build                  |
| glow          | go build                  |
|               |                           |
| git-lfs       | makefile                  |
| skopeo        | makefile                  |
| umoci         | makefile                  |
| singularity   | makefile                  |
| fzf           | makefile                  |
| direnv        | makefile, w/ weird custom logic |